pkg/controller/framework/shared_informer.go

```diff
@@ -0,0 +1,243 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package framework
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"k8s.io/kubernetes/pkg/client/cache"
+	"k8s.io/kubernetes/pkg/runtime"
+	utilruntime "k8s.io/kubernetes/pkg/util/runtime"
+)
+
+// if you use this, there is one behavior change.  When you receive a notification, the cache will be AT LEAST as
+// current as the notification, but it MAY be more current.  So if there was a create, followed by a delete, the cache
+// would NOT have your item.  All controllers should already be deduping on their side (if you forget to brush your teeth
+// at bedtime, you don't do it twice in the morning), but I doubt many people have considered the problem
+// This has advantages over the broadcaster since it allows us to share a common cache across many controllers.
+// Extending the broadcaster would have required us keep duplicate caches for each watch.
+type SharedInformer interface {
+	AddEventHandler(handler ResourceEventHandler) error
+	GetStore() cache.Store
+	// GetController gives back a synthetic interface that "votes" to start the informer
+	GetController() ControllerInterface
+	Run(stopCh <-chan struct{})
+	HasSynced() bool
+}
+
+type SharedIndexInformer interface {
+	SharedInformer
+
+	AddIndexer(indexer cache.Indexer) error
+	GetIndexer() cache.Indexer
+}
+
+func NewSharedInformer(lw cache.ListerWatcher, objType runtime.Object, resyncPeriod time.Duration) SharedInformer {
```

Comment:
> I think it would be nice to have a registry of these that we can use to look them up from so we can create new ones only if needed, etc.

I agree, but I see that as another layer built in a later pull after we've taken the easy bits in the controller manager